File: phases/09-computer-networks/18-load-balancers-l4-vs-l7-algorithms/code/main.py
"""Load Balancer Simulation — L4 vs L7, Algorithms."""

import hashlib
import random
from collections import Counter
from dataclasses import dataclass, field
from typing import List, Optional, Tuple


@dataclass
class Backend:
    addr: str
    weight: int = 1
    active_connections: int = 0
    healthy: bool = True


class LoadBalancer:
    def __init__(self, backends: List[Backend], algorithm: str = "round_robin"):
        self.backends = backends
        self.algorithm = algorithm
        self._rr_index = 0
        self._wrr_state = [0] * len(backends)
        self._wrr_total = sum(b.weight for b in backends)
        self._consistent_ring = self._build_ring()

    def route(self, client_ip: str) -> Optional[Backend]:
        healthy = [b for b in self.backends if b.healthy]
        if not healthy:
            return None

        if self.algorithm == "round_robin":
            return self._round_robin(healthy)
        elif self.algorithm == "weighted_round_robin":
            return self._weighted_round_robin(healthy)
        elif self.algorithm == "least_connections":
            return min(healthy, key=lambda b: b.active_connections)
        elif self.algorithm == "ip_hash":
            idx = int(hashlib.md5(client_ip.encode()).hexdigest(), 16) % len(healthy)
            return healthy[idx]
        elif self.algorithm == "consistent_hash":
            return self._consistent_hash(client_ip, healthy)
        elif self.algorithm == "random_two_choices":
            a, b = random.sample(healthy, min(2, len(healthy)))
            return a if a.active_connections <= b.active_connections else b

    def _round_robin(self, healthy: List[Backend]) -> Backend:
        for _ in range(len(self.backends)):
            idx = self._rr_index % len(self.backends)
            self._rr_index += 1
            if self.backends[idx].healthy:
                return self.backends[idx]

    def _weighted_round_robin(self, healthy: List[Backend]) -> Backend:
        while True:
            for i, b in enumerate(self.backends):
                if not b.healthy:
                    continue
                self._wrr_state[i] += b.weight
                if self._wrr_state[i] >= self._wrr_total:
                    self._wrr_state[i] -= self._wrr_total
                    return b

    def _consistent_hash(self, client_ip: str, healthy: List[Backend]) -> Backend:
        h = int(hashlib.md5(client_ip.encode()).hexdigest(), 16)
        for ring_hash, idx in self._consistent_ring:
            if h <= ring_hash and self.backends[idx].healthy:
                return self.backends[idx]
        for _, idx in self._consistent_ring:
            if self.backends[idx].healthy:
                return self.backends[idx]

    def _build_ring(self) -> List[Tuple[int, int]]:
        ring = []
        for i, b in enumerate(self.backends):
            for vnode in range(150):
                key = f"{b.addr}:{vnode}"
                h = int(hashlib.md5(key.encode()).hexdigest(), 16)
                ring.append((h, i))
        ring.sort()
        return ring

    def simulate_requests(self, n: int, skew: bool = False) -> Counter:
        counts = Counter()
        for i in range(n):
            if skew:
                if random.random() < 0.8:
                    client = f"10.0.0.{random.randint(1, 20)}"
                else:
                    client = f"10.0.0.{random.randint(21, 200)}"
            else:
                client = f"10.0.0.{random.randint(1, 200)}"

            backend = self.route(client)
            if backend:
                counts[backend.addr] += 1
        return counts


def main() -> None:
    print("=" * 60)
    print("Load Balancers — L4 vs L7, Algorithms")
    print("=" * 60)

    def make_backends():
        return [
            Backend("10.0.0.1:8080"),
            Backend("10.0.0.2:8080"),
            Backend("10.0.0.3:8080"),
            Backend("10.0.0.4:8080"),
        ]

    algorithms = [
        "round_robin",
        "weighted_round_robin",
        "least_connections",
        "ip_hash",
        "consistent_hash",
        "random_two_choices",
    ]

    for algo in algorithms:
        backends = make_backends()
        lb = LoadBalancer(backends, algorithm=algo)
        counts = lb.simulate_requests(1000)

        print(f"\n--- {algo} (uniform traffic) ---")
        total = sum(counts.values())
        for addr in sorted(counts.keys()):
            pct = counts[addr] / total * 100
            print(f"  {addr}: {counts[addr]:>4} ({pct:.1f}%)")

    # Skewed traffic
    print("\n--- Round Robin with 80/20 skewed traffic ---")
    backends = make_backends()
    lb = LoadBalancer(backends, algorithm="round_robin")
    counts = lb.simulate_requests(1000, skew=True)
    total = sum(counts.values())
    for addr in sorted(counts.keys()):
        pct = counts[addr] / total * 100
        print(f"  {addr}: {counts[addr]:>4} ({pct:.1f}%)")

    # Health check failure
    print("\n--- Health Check Failure Simulation ---")
    backends = make_backends()
    backends[1].healthy = False
    lb = LoadBalancer(backends, algorithm="round_robin")
    counts = lb.simulate_requests(100)
    print(f"  Backend 2 marked DOWN. Distribution over 100 requests:")
    for addr in sorted(counts.keys()):
        print(f"  {addr}: {counts[addr]}")

    # Weighted round-robin
    print("\n--- Weighted Round Robin (weights: 5, 3, 2, 1) ---")
    backends = [
        Backend("10.0.0.1:8080", weight=5),
        Backend("10.0.0.2:8080", weight=3),
        Backend("10.0.0.3:8080", weight=2),
        Backend("10.0.0.4:8080", weight=1),
    ]
    lb = LoadBalancer(backends, algorithm="weighted_round_robin")
    counts = lb.simulate_requests(1100)
    total = sum(counts.values())
    for addr in sorted(counts.keys()):
        pct = counts[addr] / total * 100
        print(f"  {addr}: {counts[addr]:>4} ({pct:.1f}%)")


if __name__ == "__main__":
    main()

File: phases/09-computer-networks/18-load-balancers-l4-vs-l7-algorithms/code/main.rs
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};
use std::net::SocketAddr;

#[derive(Debug, Clone)]
pub struct Backend {
    pub addr: SocketAddr,
    pub weight: u32,
    pub active_connections: u32,
    pub healthy: bool,
}

impl Backend {
    pub fn new(addr: &str, weight: u32) -> Self {
        Backend {
            addr: addr.parse().expect("invalid address"),
            weight,
            active_connections: 0,
            healthy: true,
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub enum Algorithm {
    RoundRobin,
    WeightedRoundRobin,
    LeastConnections,
    IpHash,
    ConsistentHash,
    RandomTwoChoices,
}

pub struct LoadBalancer {
    backends: Vec<Backend>,
    algorithm: Algorithm,
    rr_index: usize,
    wrr_current_weights: Vec<u32>,
    wrr_total_weight: u32,
    consistent_ring: Vec<(u64, usize)>,
}

impl LoadBalancer {
    pub fn new(backends: Vec<Backend>, algorithm: Algorithm) -> Self {
        let wrr_total: u32 = backends.iter().map(|b| b.weight).sum();
        let n = backends.len();
        let mut lb = LoadBalancer {
            backends,
            algorithm,
            rr_index: 0,
            wrr_current_weights: vec![0; n],
            wrr_total_weight: wrr_total,
            consistent_ring: Vec::new(),
        };
        lb.build_consistent_ring();
        lb
    }

    pub fn route(&mut self, client_ip: &str) -> Option<&Backend> {
        let healthy_count = self.backends.iter().filter(|b| b.healthy).count();
        if healthy_count == 0 {
            return None;
        }

        match self.algorithm {
            Algorithm::RoundRobin => self.round_robin(),
            Algorithm::WeightedRoundRobin => self.weighted_round_robin(),
            Algorithm::LeastConnections => self.least_connections(),
            Algorithm::IpHash => self.ip_hash(client_ip),
            Algorithm::ConsistentHash => self.consistent_hash(client_ip),
            Algorithm::RandomTwoChoices => self.random_two_choices(),
        }
    }

    fn round_robin(&mut self) -> Option<&Backend> {
        let n = self.backends.len();
        for _ in 0..n {
            let idx = self.rr_index % n;
            self.rr_index += 1;
            if self.backends[idx].healthy {
                return Some(&self.backends[idx]);
            }
        }
        None
    }

    fn weighted_round_robin(&mut self) -> Option<&Backend> {
        loop {
            for i in 0..self.backends.len() {
                if !self.backends[i].healthy {
                    continue;
                }
                self.wrr_current_weights[i] += self.backends[i].weight;
                if self.wrr_current_weights[i] >= self.wrr_total_weight {
                    self.wrr_current_weights[i] -= self.wrr_total_weight;
                    return Some(&self.backends[i]);
                }
            }
        }
    }

    fn least_connections(&self) -> Option<&Backend> {
        self.backends.iter()
            .filter(|b| b.healthy)
            .min_by_key(|b| b.active_connections)
    }

    fn ip_hash(&self, client_ip: &str) -> Option<&Backend> {
        let mut hasher = DefaultHasher::new();
        client_ip.hash(&mut hasher);
        let hash = hasher.finish() as usize;
        let healthy: Vec<&Backend> = self.backends.iter().filter(|b| b.healthy).collect();
        if healthy.is_empty() {
            return None;
        }
        Some(healthy[hash % healthy.len()])
    }

    fn consistent_hash(&self, client_ip: &str) -> Option<&Backend> {
        let mut hasher = DefaultHasher::new();
        client_ip.hash(&mut hasher);
        let hash = hasher.finish();

        for (ring_hash, idx) in &self.consistent_ring {
            if hash <= *ring_hash && self.backends[*idx].healthy {
                return Some(&self.backends[*idx]);
            }
        }
        for (_, idx) in &self.consistent_ring {
            if self.backends[*idx].healthy {
                return Some(&self.backends[*idx]);
            }
        }
        None
    }

    fn random_two_choices(&self) -> Option<&Backend> {
        use std::time::{SystemTime, UNIX_EPOCH};
        let t = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos();
        let healthy: Vec<(usize, &Backend)> = self.backends.iter()
            .enumerate()
            .filter(|(_, b)| b.healthy)
            .collect();
        if healthy.is_empty() {
            return None;
        }
        let n = healthy.len();
        let a = (t as usize) % n;
        let b = ((t >> 32) as usize) % n;
        let pick = if healthy[a].1.active_connections <= healthy[b].1.active_connections {
            a
        } else {
            b
        };
        Some(healthy[pick].1)
    }

    fn build_consistent_ring(&mut self) {
        self.consistent_ring.clear();
        for (i, backend) in self.backends.iter().enumerate() {
            for vnode in 0..150 {
                let mut hasher = DefaultHasher::new();
                format!("{}:{}", backend.addr, vnode).hash(&mut hasher);
                self.consistent_ring.push((hasher.finish(), i));
            }
        }
        self.consistent_ring.sort_by_key(|(h, _)| *h);
    }

    pub fn health_check(&mut self) {
        for backend in &mut self.backends {
            backend.healthy = true;
        }
    }

    pub fn set_unhealthy(&mut self, idx: usize) {
        if idx < self.backends.len() {
            self.backends[idx].healthy = false;
        }
    }
}

fn main() {
    println!("=== Load Balancer -- L4 vs L7, Algorithms ===\n");

    let backends = vec![
        Backend::new("10.0.0.1:8080", 1),
        Backend::new("10.0.0.2:8080", 1),
        Backend::new("10.0.0.3:8080", 1),
        Backend::new("10.0.0.4:8080", 1),
    ];

    let algorithms = [
        Algorithm::RoundRobin,
        Algorithm::WeightedRoundRobin,
        Algorithm::LeastConnections,
        Algorithm::IpHash,
        Algorithm::ConsistentHash,
        Algorithm::RandomTwoChoices,
    ];

    for algo in algorithms {
        let mut lb = LoadBalancer::new(backends.clone(), algo);
        let mut counts: [u32; 4] = [0; 4];
        let num_requests = 1000;

        for i in 0..num_requests {
            let client = format!("192.168.1.{}", i % 200);
            if let Some(backend) = lb.route(&client) {
                let ip_str = backend.addr.ip().to_string();
                let last = ip_str.split('.').last().unwrap();
                let idx: usize = last.parse::<usize>().unwrap() - 1;
                if idx < 4 {
                    counts[idx] += 1;
                }
            }
        }

        println!("{:?}:", algo);
        for (i, count) in counts.iter().enumerate() {
            let pct = (*count as f64 / num_requests as f64) * 100.0;
            println!("  Backend {} : {:>4} ({:.1}%)", i + 1, count, pct);
        }
        println!();
    }

    // Health check failure demo
    println!("--- Health Check Failure ---");
    let mut lb = LoadBalancer::new(backends.clone(), Algorithm::RoundRobin);
    lb.set_unhealthy(1);
    let mut counts: [u32; 4] = [0; 4];
    for i in 0..12 {
        let client = format!("10.0.0.{}", i);
        if let Some(backend) = lb.route(&client) {
            let ip_str = backend.addr.ip().to_string();
            let last = ip_str.split('.').last().unwrap();
            let idx: usize = last.parse::<usize>().unwrap() - 1;
            if idx < 4 {
                counts[idx] += 1;
            }
        }
    }
    println!("With backend 2 down:");
    for (i, count) in counts.iter().enumerate() {
        let status = if i == 1 { "DOWN" } else { "healthy" };
        println!("  Backend {} ({}) : {} requests", i + 1, status, count);
    }
}
